src/nginx/grpc_server_call.cc

```diff
@@ -691,6 +694,37 @@ bool NgxEspGrpcServerCall::TryReadDownstreamMessage() {
   return true;
 }
 
+// Allocate the chain link and buffer.
+ngx_chain_t *NgxEspGrpcServerCall::AllocNgxBufChain(size_t buflen) {
+  ngx_chain_t *cl = ngx_chain_get_free_buf(r_->pool, &buf_free_);
+  if (!cl) {
+    return nullptr;
+  }
+  ngx_buf_t *buf = cl->buf;
+
+  if (buflen > size_t(buf->end - buf->start)) {
+    // Abandon the small buf, allocate a bigger one.
```

Comment:
What happens to the small buffer? Is it left unused and deallocated when the request finalizes?
If so, should we try to use it by adding it back to the free list?

Reply:
the smaller buffers are dropped,  they will be freed when the whole request (grpc stream) is done. 
Since we only re-cycle one buffer so we only keep the largest one in the free list.  If we put the smaller ones back to free list,  they could be used, but their size maybe too small so putting them back to free list is troublesome.